File: src/endpoints/checklist/get_index.py
from src import output
from src.libs.db import DB, Table
from src.modules.verify import checking_token
from src.modules.mod import MOD
from . import __types__ as types
from ..tag import __libs__ as tag_libs

async def get_index(params: dict = {}, req = None, _db: DB = None, _check_token = True):

    # set values
    result = None
    db = _db if _db else DB().connect()

    try:
        # set params
        params = types.GetIndex(**params)

        # checking token
        if _check_token: checking_token(req, db)

        # set fields
        fields = params.fields.split(',') if params.fields else None

        # set mod
        mod = MOD(params.mod or '')

        # set data assets
        where = []
        values = {}
        join = []

        # set base params
        if params.content:
            where.append(f'and content LIKE "%{params.content}%"')
        if params.start and params.end:
            where.append(f'and (`created_at` BETWEEN "{params.start} 00:00:00" AND "{params.end} 23:59:59")')

        # set tag
        if params.tag:
            from ..tag import __libs__ as tag_lib
            _tags = ','.join(params.tag.split(','))
            where.append(f'AND checklist.srl IN (SELECT map_tag.module_srl FROM map_tag WHERE map_tag.module LIKE :module AND map_tag.tag_srl IN ({_tags}))')
            values['module'] = tag_lib.Module.CHECKLIST

        # get total
        total = db.get_count(
            table_name=Table.CHECKLIST.value,
            where=where,
            join=join,
            values=values,
        )
        if not (total > 0): raise Exception('No data', 204)

        # get index
        index = db.get_items(
            table_name=Table.CHECKLIST.value,
            fields=fields,
            where=where,
            join=join,
            limit={ 'size': params.size, 'page': params.page },
            order={ 'order': params.order, 'sort': params.sort },
            values=values,
            unlimited=params.unlimited,
        )
        def transform_item(item: dict) -> dict:
            # MOD / tag
            if mod.check('tag'):
                item['tag'] = tag_libs.get_index(
                    _db=db,
                    module=tag_libs.Module.CHECKLIST,
                    module_srl=item.get('srl'),
                )
            return item
        index = [ transform_item(item) for item in index ]

        # set result
        result = output.success({
            'message': 'Complete get checklist index.',
            'data': {
                'total': total,
                'index': index,
            },
        }, _req=req)
    except Exception as e:
        result = output.exc(e, _req=req)
    finally:
        if not _db and db: db.disconnect()
        return result

File: src/routes/checklist/get-index.ts
import DB, { db } from '@/classes/DB'
import ServiceError from '@/classes/ServiceError'
import Service from '@/classes/Service'
import MOD from '@/classes/MOD'
import * as tagHelper from '@/routes/tag/__helper'
import type { ChecklistModel } from './__model'

type GetIndexParams = {
  query: ChecklistModel['getIndexQuery']
  service: Service
}

export default async function getIndex({ query, service }: GetIndexParams)
{
  try
  {
    // set assets
    const _table = `${DB.TABLE.CHECKLIST} AS c`
    const _field = query.field ? query.field.split(',') : ''
    let _where: string[] = []
    let _values: ZZ = {}
    let _join: string[] = []
    if (query.content)
    {
      _where.push(`AND content LIKE '%' || $content || '%'`)
      _values['$content'] = query.content.trim()
    }
    if (query.start && query.end)
    {
      _where.push(`AND (created_at BETWEEN $dateStart AND $dateEnd)`)
      _values['$dateStart'] = `${query.start} 00:00:00`
      _values['$dateEnd'] = `${query.end} 23:59:59`
    }
    if (query.tag)
    {
      const _tags = query.tag.split(',').join(',')
      _where.push(`AND c.srl IN (SELECT mt.module_srl FROM ${DB.TABLE.MAP_TAG} AS mt WHERE mt.module LIKE $tag_module AND mt.tag_srl in (${_tags}))`)
      _values['$tag_module'] = tagHelper.MODULE.CHECKLIST
    }

    // get count
    const count = db.getCount({
      table: _table,
      where: _where,
      join: _join,
      values: _values,
    })
    if (count.data <= 0) throw new ServiceError('No data', { status: 204 })

    // get index data
    const index = db.getIndex({
      table: _table,
      prefix: 'DISTINCT',
      field: _field,
      where: _where,
      join: _join,
      order: query.order,
      page: query.page ?? 1,
      size: query.size ?? service.preference['checklist.index.size'],
      values: _values,
    })

    // set MOD
    const _mod: MOD = new MOD(query.mod)

    // 인덱스 데이터 컨버팅
    const _index = index.data.map((o: ZZ) => {
      // MOD / tag
      if (_mod.check('tag'))
      {
        const _index = db.getIndex({
          table: `${DB.TABLE.TAG} AS t`,
          field: 't.srl,t.name',
          join: `JOIN ${DB.TABLE.MAP_TAG} AS mt ON mt.tag_srl = t.srl`,
          where: [
            'AND module LIKE $module',
            'AND module_srl = $module_srl'
          ],
          values: {
            '$module': tagHelper.MODULE.CHECKLIST,
            '$module_srl': o.srl,
          },
        })
        o.tag = _index.data?.length > 0 ? _index.data : []
      }
      return o
    })

    return {
      total: count.data,
      index: _index,
    }
  }
  catch (_e: any)
  {
    throw new ServiceError('Failed to get Checklist index.', {
      status: _e.status,
      text: _e.message,
      cause: _e,
    })
  }
}
